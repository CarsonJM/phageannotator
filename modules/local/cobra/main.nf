process COBRA {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/cobra-meta:1.2.3--pyhdfd78af_0':
        'biocontainers/cobra-meta:1.2.3--pyhdfd78af_0' }"

    input:
    tuple val(meta), path(fasta), path(coverage), path(viral_contigs), path(bam)
    val assembler
    val mink
    val maxk

    output:
    tuple val(meta), path("${prefix}/COBRA_extended.fasta.gz")  , emit: extended_assemblies
    tuple val(meta), path("${prefix}/COBRA_joining_summary.txt"), emit: joining_summary
    path "versions.yml"                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    cobra-meta \\
        --fasta ${fasta} \\
        --coverage ${coverage} \\
        --query ${viral_contigs} \\
        --mapping ${bam} \\
        --assembler ${assembler} \\
        --mink ${mink} \\
        --maxk ${maxk} \\
        --threads ${task.cpus} \\
        --output ${prefix} \\
        $args

    # if the output is empty, use input file
    if [ -e ${prefix}/COBRA_extended.fasta ]
    then
        gzip ${prefix}/*.fasta
        cat ${prefix}/*fasta.gz > ${prefix}/COBRA_extended.fasta.gz
    else
        gzip -c ${fasta} > ${prefix}/COBRA_extended.fasta
        touch ${prefix}/COBRA_joining_summary.txt
    fi

    gzip ${prefix}/*.fasta

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cobra: \$(echo \$(cobra-meta --version 2>&1) | sed 's/^.*cobra v//' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir ${prefix}
    touch ${prefix}/COBRA_extended.fasta
    gzip ${prefix}/COBRA_extended.fasta
    touch ${prefix}/COBRA_joining_summary.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cobra: \$(echo \$(cobra-meta --version 2>&1) | sed 's/^.*cobra v//' ))
    END_VERSIONS
    """
}
