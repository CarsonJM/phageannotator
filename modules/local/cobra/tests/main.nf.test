nextflow_process {

    name "Test process: COBRA"
    script "../main.nf"
    process "COBRA"
    tag "modules"
    tag "modules_local"
    tag "cobra"


    test("contigs.fasta.gz & coverage.tsv & viral_contigs & bam") {

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fasta/contigs.fasta", checkifExists: true),
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/test_coverage.tsv", checkifExists: true),
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/viral_contigs.tsv", checkifExists: true),
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/test.sorted.bam", checkifExists: true)
                        ]
                )
                input[1] = 'metaspades'
                input[2] = 27
                input[3] = 127
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("contigs.fasta.gz & coverage.tsv & viral_contigs & bam - stub") {

        options "-stub"

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fasta/contigs.fasta", checkifExists: true),
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/test_coverage.tsv", checkifExists: true),
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/viral_contigs.tsv", checkifExists: true),
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/test.sorted.bam", checkifExists: true)
                        ]
                )
                input[1] = 'metaspades'
                input[2] = 27
                input[3] = 127
                """
            }
        }

        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
