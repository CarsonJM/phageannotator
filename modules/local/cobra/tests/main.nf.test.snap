{
    "contigs.fasta.gz & coverage.tsv & viral_contigs & bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_extended.fasta.gz:md5,0b52beff348887e1552c9c956ba53020"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_joining_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,3412a0d78ed4f7a64e348cc63b3a86a3"
                ],
                "extended_assemblies": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_extended.fasta.gz:md5,0b52beff348887e1552c9c956ba53020"
                    ]
                ],
                "joining_summary": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_joining_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3412a0d78ed4f7a64e348cc63b3a86a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-28T11:54:47.803389179"
    },
    "contigs.fasta.gz & coverage.tsv & viral_contigs & bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_extended.fasta.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_joining_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,3412a0d78ed4f7a64e348cc63b3a86a3"
                ],
                "extended_assemblies": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_extended.fasta.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "joining_summary": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_joining_summary.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3412a0d78ed4f7a64e348cc63b3a86a3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-28T11:54:57.388744544"
    }
}